Add support for ListenBrainz compatible servers (like Maloja)

subjut opened the pull request:
Add Preference to set server URL. Clicking reset button on page will reset the URL to the default ListenBrainz URL. Change the playback menu entry according to which URL is used. Refactor to check token with `is_valid_token()`,  which also accounts for [Maloja](https://github.com/krateng/maloja) token length. Update `README.md` with custom server information.

I have tested both the default ListenBrainz and my own Maloja server instance and it seems to work without a problem for me. But a thorough review of the code is probably advisable. It's been a while since I did anything in C++…  ;-)

subjut committed:
Add support for ListenBrainz compatible servers

Add Preference to set server URL. Clicking reset button on page
will reset the URL to the default ListenBrainz URL. Change the
playback menu entry according to which URL is used. Refactor
to check token with is_valid_token() that also accounts for Maloja
token length. Update README with custom server information.

subjut committed:
Fix typo

subjut committed:
Update preference pane picture

subjut committed:
Merge branch 'main' of https://github.com/subjut/foo_listenbrainz2

phw commented on src/foo_listenbrainz2.h:
```diff
@@ -47,6 +47,15 @@ namespace lbz
+	{
+		if (mbid == nullptr) return false;
+		// 62 is the token length of Maloja. There's probably a better way to validate the token.
+		if (is_uuid(mbid) || (strlen(mbid) == 62 )) return true;
```
It's been a while since I set up Maloja, but wasn't the token configured in a config file and could be set freely?

subjut replied:
Hmm. That could be. I configured it through the web interface and the token defaults to 62 chars there. I haven't tested setting it in the config file.
So this needs some changes. Like you suggested above, it should be checked for being a valid uuid for ListenBrainz, or simply any non-empty string for any custom server. I can amend the code tomorrow or so.

subjut replied:
I also thought about adding a checkbox to enable and disable the custom API URL field in the settings, like the query field. That would also simplify the check for ListenBrainz vs. custom server, but complicate the preference pane logic a bit. But it would probably be a little more user friendly.

phw replied:
Ah yes, there is a UI for this now. When I set up my instance the first time there was just that config file and I set any random string.

Agreed, let's do it like you say and just check for non-empty string in case it's not ListenBrainz

subjut replied:
I refactored the token and API URL check functions (see commit [5cc3739](https://github.com/phw/foo_listenbrainz2/pull/18/commits/5cc37395a2cfae0b1768117dce0ed9797aa37915)) and they seem to work correctly.

phw commented on src/foo_listenbrainz2.h:
```diff
@@ -47,6 +47,15 @@ namespace lbz
 		return regex_search(mbid, rx);
 	}
 
+	static bool is_valid_token(const char* mbid)
```
I'd not name the parameter `mbid`, that confused me at first. I would expect a MusicBrainz UUID. Let's just call it e.g. `token`.

subjut replied:
Sure!

phw reviewed (changes requested):
Very cool, thanks a lot. I'll give this a test run with my Maloja instance as soon as I'm back at my laptop.

Just some minor comments.

Apart of this the only thing to that bugs me a bit is the now a bit vague token validation (and the many places where it gets performed). We probably could do the validation depending on the server URL, but maybe that will become rather messy overall. So I'd be fine merging this generally as it is.

subjut committed:
Rename token validation function parameter

subjut committed:
Improve token and API URL checks

Move both check functions to preferences. Make check for valid
token lenient when using non-default ListenBrainz server: string
has to be longer than 0.
Align preference dialog elements.

phw commented on src/preferences.h:
```diff
@@ -5,18 +5,54 @@ namespace lbz
 	}
+
+	// Returns true, if the API URL is the default ListenBrainz service.
+	static bool is_listenbrainz()
```
Just because my C++ is also a bit rusty, but isn't the use of `static` for a function inside a header essentially the same as `inline`? Or is there some other difference?

subjut replied:
I honestly don't know. Since the `is_uuid()` function is defined the same way, I just copied it.

Edit: [This](https://gist.github.com/htfy96/50308afc11678d2e3766a36aa60d5f75) might clarify some things. I don't pretend to exactly understand that, but since there is only one namespace `lbz`, the static functions should only be present once. I don't think it matters much here, though.

phw replied:
Good. Would be the same here :)

I *think* defining it as inline would have more or less the same effect. But really, it likely makes no difference here at all for the use case.

phw reviewed (approved):
Tested it against my maloja instance. This looks and works great. The rework with the new token validation also makes the code significantly clearer. Btw, I can also confirm variable size Maloja tokens, I have several with varying sizes defined (for whatever reason, but I created those before there was the UI for it).

Thanks a lot, I'll merge it
